### DISCUSSION TITLE
How does gtk-osx-setup.sh fetch modulesets?

### DISCUSSION THREAD

**Discussion by cpatulea** [Xpra-org/xpra#4091] [Q&A]
I'm trying to build for Mac OS using these instructions: https://github.com/Xpra-org/xpra/blob/master/docs/Build/MacOS.md

The largest part is the `gtk-osx-setup.sh`, this question is actually entirely about gtk-osx-build part.

In the repository, `gtk-osx-setup.sh` fetches the jhbuild custom file from gitlab.gnome.org/GNOME/gtk-osx/raw/master: https://github.com/Xpra-org/gtk-osx-build/blob/fb9ea7cb7f00b8f4846d3aaff7c0deda064804ae/gtk-osx-setup.sh#L254C18-L254C59

That seems to suggest it uses the modulesets (package URLs, versions, etc) from upstream, not from github.com/Xpra-org/gtk-osx-build.

For example, upstream uses libpng-1.6.40 (https://gitlab.gnome.org/GNOME/gtk-osx/-/blob/master/modulesets-stable/gtk-osx-bootstrap.modules?ref_type=heads#L39) while Xpra-org/gtk-osx-build uses libpng-1.6.39 (https://github.com/Xpra-org/gtk-osx-build/blob/dc68bb3f850d5f939992718d0dbd195cd970f976/modulesets/gtk-osx-bootstrap.modules#L21).

And indeed, it seems the build download 1.6.40 (upstream): https://github.com/cpatulea/gtk-osx-build/actions/runs/7391404577/job/20108152043 (search for "libpng-").

The build fails for some other unrelated issue, but I'm trying to understand where is the source of modulesets first (so I can tweak URLs, etc).

Context: https://github.com/Xpra-org/gtk-osx-build/issues/27

**Comment by cpatulea** [comment]
Ahh, I missed the key step in the instructions:

```
Configure `jhbuild` to use our modules
curl -o ~/.config/jhbuildrc-custom \
     https://raw.githubusercontent.com/Xpra-org/gtk-osx-build/master/jhbuildrc-custom
```

Sorry about that. Let me try with that explicit step.

**Comment by cpatulea** [comment]
Hmm, still using libpng-1.6.40 (https://github.com/cpatulea/gtk-osx-build/actions/runs/7391984380/job/20109741842).

`jhbuildrc` is from upstream gtk-osx.

but I think it's supposed to source the `jhbuildrc-custom` through this line: https://gitlab.gnome.org/GNOME/gtk-osx/-/blob/master/jhbuildrc-gtk-osx?ref_type=heads#L474

**Comment by cpatulea** [comment]
Argh, I was looking at the wrong path (`modulesets` vs `modulesets-stable`): both upstream and Xpra-org use libpng 1.6.40. I think it's actually using the correct modulesets, just now I'm stuck again on the www.cairographics.org outage.

**Comment by cpatulea** [comment]
Hmm, nope? I modified the URLs in `modulesets-stable/gtk-osx.modules`: https://github.com/cpatulea/gtk-osx-build/commit/a7426d8c4591b1dd487dcdb648c6e2161e388d19

but the build still seems to use the old URLs: https://github.com/cpatulea/gtk-osx-build/actions/runs/7392489963/job/20111034296

[edit] Oh, probably because it take the modulesets from Xpra-org/gtk-osx-build head:
```
moduleset="https://raw.githubusercontent.com/Xpra-org/gtk-osx-build/master/xpra-gtk3.modules"
```
(so, not from my personal fork)

**Comment by cpatulea** [comment]
Good way to validate the `jhbuildrc-custom` locally:

```
cp -af jhbuildrc-custom ~/.config/jhbuildrc-custom && ~/.new_local/bin/jhbuild update
```

**Comment by cpatulea** [comment]
Yes! Got past `jhbuild update`. It has been on `jhbuild bootstrap-gtk-osx` for 18 minutes and counting: https://github.com/cpatulea/gtk-osx-build/actions/runs/7392978607/job/20112231399

**Comment by totaam** [comment]
You're bound to hit some of the unresolved issues with https://github.com/Xpra-org/gtk-osx-build (ie: the jpeg dylib, etc)

Even on decent native hardware, building the whole thing (not even the xpra app itself) takes quite a while.

Then there are a bunch of manual steps nowadays - perhaps not documented properly, to do with installing python packages that no longer (or even never did) support installation via jhbuild's setuptools.

It might be possible to install pandoc using scripting.
I don't think it is reasonable to build assets using the CI tools though, it takes far too long, breaks far too often and handling signed binaries is likely to be very difficult to do securely.

**Comment by totaam** [comment]
@cpatulea probably knows more about modulesets than I do now!
(and most of the issues above have been resolved too)
